[DLv2] Make graph `traverse` working with unhashable `DataPipe`

ejguan opened the pull request:
Summary:
This Diff removes the requirement for `traverse` function that `DataPipe` needs to be hash-able. `traverse` function now is using `id` of `DataPipe` instance rather than `DataPipe` itself as the key for both `cache` and graph.

But, it requires the changes of type of `DataPipeGraph` from `Dict[DataPipe, "DataPipeGraph"]` to `Dict[int, Tuple[DataPipe, "DataPipeGraph"]]`.

Differential Revision: D37354153

Ref PR in TorchData: https://github.com/pytorch/data/pull/559

ejguan commented on torch/utils/data/graph.py:
```diff
@@ -4,11 +4,13 @@
-__all__ = ["traverse", ]
+__all__ = ["DataPipe", "DataPipeGraph", "traverse"]
 
 DataPipe = Union[IterDataPipe, MapDataPipe]
```
Should I rename it to `DataPipeType`?

NivekT replied:
I am indifferent.

NivekT commented on torch/utils/data/graph.py:
```diff
@@ -70,21 +72,22 @@ def reduce_hook(obj):
 
-    cache.add(datapipe)
+    dp_id = id(datapipe)
+    cache.add(id(dp_id))
```
```suggestion
    cache.add(dp_id)
```
I think this is what you want and should fix the infinite recursion in CI.

NivekT commented on torch/utils/data/graph.py:
```diff
@@ -4,11 +4,13 @@
-__all__ = ["traverse", ]
+__all__ = ["DataPipe", "DataPipeGraph", "traverse"]
 
 DataPipe = Union[IterDataPipe, MapDataPipe]
+DataPipeGraph = Dict[int, Tuple[DataPipe, "DataPipeGraph"]]  # type: ignore[misc]
```
Do their `__module__` need to change or they need to be added to `__init__` to fix the following CI error?
https://github.com/pytorch/pytorch/runs/7114135518?check_suite_focus=true

NivekT reviewed (commented):
Mostly LGTM! Needs to fix the CI issues

ejguan committed:
[DLv2] Make graph `traverse` working with unhashable `DataPipe` (#80509)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/80509

X-link: https://github.com/pytorch/data/pull/559

This Diff removes the requirement for `traverse` function that `DataPipe` needs to be hash-able. `traverse` function now is using `id` of `DataPipe` instance rather than `DataPipe` itself as the key for both `cache` and graph.

But, it requires the changes of type of `DataPipeGraph` from `Dict[DataPipe, "DataPipeGraph"]` to `Dict[int, Tuple[DataPipe, "DataPipeGraph"]]`.

Test Plan:
`buck test //caffe2/test:datapipe`
`buck test //pytorch/data/test:tests`
`buck test //pytorch/data/fb/tests:disaggregated_dpp_reading_service_test`
`buck test //pytorch/data/fb/tests:dpp_reading_service_utils_test`

Reviewed By: Miiira

Differential Revision: D37354153

fbshipit-source-id: 7e19a85d22af0b7c339fa84e51d1ec44fd8f6b8d
